add validation of update mirror urls

christopherpross opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Fixes #17205 

### Summary of the issue:
Users could configure an invalid update mirror URL, which would only be discovered when attempting to check for updates. This PR implements a validation mechanism that ensures the specified update mirror is valid before allowing it to be set in the settings.

### Description of user facing changes
A new validation process has been added when setting an update mirror URL in NVDA's settings. Users will now receive feedback if the URL they provide is not a valid update mirror. The "Test" button in the settings will now ensure that the mirror responds with the expected format, preventing invalid configurations.

### Description of development approach
- Refactored parsing logic for update responses into a new function: `parseUpdateCheckResponse`.
- Defined the minimum schema for an update mirror response based on the following required keys:
  - `version`
  - `launcherUrl`
  - `apiVersion`
- Implemented a new function `_isResponseUpdateMirrorValid` in `settingsDialogs.py`, which calls `parseUpdateCheckResponse` to validate the mirror's response.
- Added `_isResponseUpdateMirrorValid` as the `responseValidator` in the `_SetURLDialog` for update mirrors.

### Testing strategy:
- Ran NVDA from source.
- Ensured the update mirror URL was set to "no Mirror".
- Test 1: Set the URL to "https://www.nvaccess.org/nvdaUpdateCheck".
  - Pressed the "Test" button and verified that the URL was marked as valid.
- Test 2: Set the URL to "https://google.de".
  - Pressed the "Test" button and verified that the URL was marked as invalid.
- Test 3: Set the URL to "https://github.com".
  - Pressed the "Test" button and verified that the URL was marked as invalid.
- Test 4: Set the URL to the Chinese NVDA Community Update Mirror (https://nvaccess.mirror.nvdadr.com/nvdaUpdateCheck).
  - Verified that this URL was marked as valid.
- Additional tests with random strings to ensure that invalid URLs are correctly marked as invalid.


### Known issues with pull request:
No known issues.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

christopherpross committed:
refactor(updateCheck): extract response parsing logic into separate function (#17205)

Moved the logic for parsing update check responses into a new function `parseUpdateCheckResponse`.

christopherpross committed:
add(updateCheck): add update mirror url validation to the UI (#17205)

pre-commit-ci[bot] committed:
Pre-commit auto-fix

christopherpross committed:
documentation(updateCheck): update changelog to mention the pull request. (#17205)

christopherpross committed:
Merge branch 'feature/#17205' of https://github.com/christopherpross/nvda into feature/#17205

seanbudd commented on source/updateCheck.py:
```diff
@@ -118,6 +118,27 @@ def getQualifiedDriverClassNameForStats(cls):
 	return "%s (core)" % name
 
 
+def parseUpdateCheckResponse(data: str) -> Optional[Dict[str, str]]:
```
```suggestion
def parseUpdateCheckResponse(data: str) -> dict[str, str] | None:
```

christopherpross committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

SaschaCowley reviewed (changes requested):
Thanks @christopherpross. A couple of minor adjustments, plus I think you may have forgotten to push the schema validation.

christopherpross committed:
Update user_docs/en/changes.md

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

christopherpross commented on source/gui/settingsDialogs.py:
```diff
@@ -5595,3 +5596,18 @@ def _isResponseAddonStoreCacheHash(response: requests.Response) -> bool:
 	return isinstance(data, str) and bool(data)
+
+
+def _isResponseUpdateMirrorValid(response: requests.Response) -> bool:
```
@SaschaCowley  Hey, sorry I see. While merging, there is a change not applied. So this is lost. In the next few days, I will fix that. Sorry for the inconvinience.

seanbudd commented:
Hi @christopherpross - just confirming you've seen Sascha's review and intend to still work on this?

christopherpross commented:
@seanbudd  and @SaschaCowley  sorry for the delay, my real life comes in the way. I am so sorry. If you have a deathline for this fix, someone can, if they want take this and fix the remaining missing code. I has forgotten something to push and deleted my local repo. So in the next few days I will try to get into this. But I understand fully if you have a deathline and you have to fix this issue fast. Sorry for the inconventience.

christopherpross committed:
add validation function for update mirror urls and use it in the UI and the updateCheck function. #17205

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -118,15 +118,64 @@ def getQualifiedDriverClassNameForStats(cls):
 	return "%s (core)" % name
 
 
+def parseUpdateCheckResponse(data: str) -> dict[str, str] | None:
```
To make the code more self-documenting and easier to debug, could you create a new class to hold update information (perhaps as a [`dataclasses.dataclass`](https://docs.python.org/3/library/dataclasses.html) or [`typing.NamedTuple`](https://docs.python.org/3/library/typing.html#special-types))?

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -180,25 +232,22 @@ def checkForUpdate(auto: bool = False) -> Optional[Dict]:
-			# and then retry the update check.
-			log.debug(f"Fetching update data from {url}")
+			# Retry the update check
+			log.debug(f"Retrying update data fetch from {url}")
```
```suggestion
			log.debug(f"Retrying update check from {url}")
```

SaschaCowley commented on source/gui/settingsDialogs.py:
```diff
@@ -5595,3 +5596,10 @@ def _isResponseAddonStoreCacheHash(response: requests.Response) -> bool:
 	return isinstance(data, str) and bool(data)
+
+
+def _isResponseUpdateMirrorValid(response: requests.Response) -> bool:
```
Could you rename this to be a little clearer? Perhaps something like `isResponseUpdateMetadata`

SaschaCowley reviewed (changes requested):
Thanks @christopherpross. A few suggestions to improve the quality of the update check code while you're here. Please also update the copyright headers of the files you have touched to reflect that they have been modified in 2025, and add yourself to the list of copyright holders, if you like

christopherpross committed:
apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

christopherpross committed:
refactor(updateCheck): replace dict with UpdateInfo dataclass #17205

Introduced the UpdateInfo dataclass to replace the dictionary used for update metadata. This improves type safety, readability, and maintainability by providing a structured and self-documenting representation of update information.

christopherpross committed:
Adjust copyright file headers for touched files

pre-commit-ci[bot] committed:
Pre-commit auto-fix

christopherpross commented:
@SaschaCowley 
Done, I have implemented all the suggested changes.

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -118,15 +146,82 @@ def getQualifiedDriverClassNameForStats(cls):
 	return "%s (core)" % name
 
 
+def parseUpdateCheckResponse(data: str) -> Optional[UpdateInfo]:
```
Can you:
* Raise appropriate exceptions when parsing fails, and handle them gracefully when calling this function?
* Make this a classmethod of `UpdateInfo`?

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -118,15 +146,82 @@ def getQualifiedDriverClassNameForStats(cls):
+	:return: An UpdateInfo object containing the update metadata, or None if the format is invalid.
+	"""
+	# Ensure the data is valid using isValidUpdateMirrorResponse
+	if not isValidUpdateMirrorResponse(data):
```
This is duplicating work. Instead, this function should raise errors, and `isValidUpdateMirrorResponse` should try calling this class and returning `False` if it fails. When making this change, the call to `isValidUpdateMirrorResponse` will need to be removed, or you'll get infinite recursion.

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -118,15 +146,82 @@ def getQualifiedDriverClassNameForStats(cls):
+	)
+
+
+def isValidUpdateMirrorResponse(responseData: str) -> bool:
```
Re-implement per above

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -118,15 +146,82 @@ def getQualifiedDriverClassNameForStats(cls):
 
 
-def checkForUpdate(auto: bool = False) -> Optional[Dict]:
+def checkForUpdate(auto: bool = False) -> Optional[UpdateInfo]:
```
```suggestion
def checkForUpdate(auto: bool = False) -> UpdateInfo | None:
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -118,15 +146,82 @@ def getQualifiedDriverClassNameForStats(cls):
-	@return: Information about the update or C{None} if there is no update.
-	@raise RuntimeError: If there is an error checking for an update.
+
+	:param auto: Whether this is an automatic check for updates.
```
```suggestion
	:param auto: Whether this is an automatic check for updates, defaults to False.
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -180,25 +278,27 @@ def checkForUpdate(auto: bool = False) -> Optional[Dict]:
+
+	data = res.read().decode("utf-8")  # Ensure the response is decoded correctly
+
+	if not isValidUpdateMirrorResponse(data):
+		raise RuntimeError(
+			"The update response is invalid. Ensure the update mirror returns a properly formatted response."
+		)
+
```
This is tripling work
```suggestion
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -367,7 +467,7 @@ def _error(self):
 		)
 
-	def _result(self, info: Optional[Dict]) -> None:
+	def _result(self, info: Optional[UpdateInfo]) -> None:
```
```suggestion
	def _result(self, info: UpdateInfo | None) -> None:
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -424,7 +524,7 @@ class UpdateResultDialog(
 	helpId = "GeneralSettingsCheckForUpdates"
 
-	def __init__(self, parent, updateInfo: Optional[Dict], auto: bool) -> None:
+	def __init__(self, parent, updateInfo: Optional[UpdateInfo], auto: bool) -> None:
```
```suggestion
	def __init__(self, parent, updateInfo: UpdateInfo | None, auto: bool) -> None:
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -671,20 +771,22 @@ class UpdateDownloader(garbageHandler.TrackedObject):
+		Constructor for the update downloader.
+		:param updateInfo: An UpdateInfo object containing the metadata of the update,
+		including version, URLs, and compatibility information.
+		:type updateInfo: UpdateInfo
```
We don't include type info in docstrings when that type info is fully described by type annotations.
```suggestion
```

SaschaCowley reviewed (changes requested):
A few minor tweaks

christopherpross committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

christopherpross commented:
@SaschaCowley  Ok, done I have implemented all suggested changes. I have also resolved a conflict in settingsDialog.py

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -94,6 +97,66 @@
+		for line in data.splitlines():
+			try:
+				key, val = line.split(": ", 1)
+				metadata[key] = val
```
If we use a similar approach to above, we can grab all good keys, and silently drop bad keys, so we only need to unpack `metadata` to initialise `cls`. We can also remove keys from `requiredKeys` to make checking for all required keys easier later.
```suggestion
				if key in recognizedKeys:
					metadata[key] = val
					requiredKeys.difference_update(key)
				else:
					log.debug(f"Dropping unknown key {key} = {val}.")
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -94,6 +97,66 @@
+				raise ValueError(f"Invalid line format in update response: {line}")
+
+		# Check for missing required keys
+		missing_keys = [key for key in required_keys if key not in metadata]
+		if missing_keys:
```
If you decide to use the approach suggested above, this can be simplified to
```suggestion
		if requiredKeys:
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -94,6 +97,66 @@
+			raise ValueError(f"Missing required keys in update response: {', '.join(missing_keys)}")
+
+		# Create and return UpdateInfo
+		return cls(
+			# Required keys
+			version=metadata["version"],
+			launcherUrl=metadata["launcherUrl"],
+			apiVersion=metadata["apiVersion"],
+			# Optional keys
+			launcherHash=metadata.get("launcherHash", None),
+			apiCompatTo=metadata.get("apiCompatTo", None),
+			changesUrl=metadata.get("changesUrl", None),
+			launcherInteractiveUrl=metadata.get("launcherInteractiveUrl", None),
+		)
```
It would probably be easier to do
```suggestion
		return cls(**metadata)
```
The further suggestions above should help to make this fairly bullet-proof,, but if you wanted to you could wrap this in a try/except, except `TypeError`, and re-raise it as a `ValueError` since that's more appropriate here.

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -121,15 +184,31 @@ def getQualifiedDriverClassNameForStats(cls):
 
 
-def checkForUpdate(auto: bool = False) -> Optional[Dict]:
+def checkForUpdate(auto: bool = False) -> Optional[UpdateInfo]:
```
```suggestion
def checkForUpdate(auto: bool = False) -> UpdateInfo | None:
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -370,7 +451,7 @@ def _error(self):
 		)
 
-	def _result(self, info: Optional[Dict]) -> None:
+	def _result(self, info: Optional[UpdateInfo]) -> None:
```
```suggestion
	def _result(self, info: UpdateInfo | None) -> None:
```

SaschaCowley commented on source/updateCheck.py:
```diff
@@ -427,7 +508,7 @@ class UpdateResultDialog(
 	helpId = "GeneralSettingsCheckForUpdates"
 
-	def __init__(self, parent, updateInfo: Optional[Dict], auto: bool) -> None:
+	def __init__(self, parent, updateInfo: Optional[UpdateInfo], auto: bool) -> None:
```
```suggestion
	def __init__(self, parent, updateInfo: UpdateInfo | None, auto: bool) -> None:
```

SaschaCowley commented on user_docs/en/changes.md:
Please document API breaking changes (by my count, `checkForUpdate`, `UpdateResultDialog.__init__`, `UpdateDownloader.__init__`)

SaschaCowley reviewed (changes requested):
Some suggestions on how to make this more maintainable and performant, as well as on keeping to our coding standards.

christopherpross committed:
Add api breaking changes to 'changes.md'

christopherpross commented:
@SaschaCowley Ok, I have applied the suggestions and documented the breaking changes.

SaschaCowley committed:
Made `parseUpdateCheckResponse" a classmethod

SaschaCowley committed:
Fixed docstrings

SaschaCowley committed:
Fix implementation of parseUpdateCheckResponse

SaschaCowley committed:
Remove  isValidUpdateMirrorResponse and move logic directly into _isResponseUpdateMetadata

SaschaCowley committed:
Formatted

SaschaCowley committed:
Fix up changelog

SaschaCowley committed:
Merge branch 'master' into feature/#17205

# Conflicts:
#	source/gui/settingsDialogs.py
#	user_docs/en/changes.md
